components/engine_test/src/lib.rs

```diff
@@ -243,60 +253,80 @@ pub mod kv {
-            self.registry.lock().unwrap().get(&(id, suffix)).cloned()
-        }
+            let mut reg = self.registry.lock().unwrap();
+            if let Some(suffix) = suffix {
+                if let Some(tablet) = reg.get(&(id, suffix)) {
+                    // Target tablet exist in the cache
+
+                    if options.create_new() {
+                        return Err(box_err!(
+                            "region {} {} already exists",
+                            id,
+                            tablet.as_inner().path()
+                        ));
+                    }
+                    return Ok(tablet.clone());
+                } else if !options.cache_only() {
+                    let tablet_path = self.tablet_path(id, suffix);
+                    let tablet = self.open_tablet_raw(&tablet_path, id, suffix, options.clone())?;
+                    if !options.skip_cache() {
+                        reg.insert((id, suffix), tablet.clone());
+                    }
+                    return Ok(tablet);
+                }
+            } else if options.cache_only() {
+                // This branch reads an arbitrary tablet with region id `id`
 
-        fn open_tablet_cache_any(&self, id: u64) -> Option<KvTestEngine> {
-            let reg = self.registry.lock().unwrap();
-            if let Some(k) = reg.keys().find(|k| k.0 == id) {
-                return Some(reg.get(k).unwrap().clone());
+                if let Some(k) = reg.keys().find(|k| k.0 == id) {
+                    return Ok(reg.get(k).unwrap().clone());
+                }
             }
-            None
-        }
 
-        fn open_tablet_raw(&self, path: &Path, _readonly: bool) -> Result<KvTestEngine> {
-            if !KvTestEngine::exists(path.to_str().unwrap_or_default()) {
+            Err(box_err!(
+                "tablet with region id {} suffix {:?} does not exist",
+                id,
+                suffix
+            ))
+        }
+
+        fn open_tablet_raw(
+            &self,
+            path: &Path,
+            _id: u64,
+            _suffix: u64,
+            options: OpenOptions,
+        ) -> Result<KvTestEngine> {
+            // Even though neither options.create nor options.create_new are true, if the
+            // tablet files already exists, we will open it by calling
+            // inner.create_tablet. In this case, the tablet exists but not in the cache
+            // (registry).
+            if !options.create()
+                && !options.create_new()
+                && !KvTestEngine::exists(path.to_str().unwrap_or_default())
+            {
                 return Err(box_err!(
                     "path {} does not have db",
                     path.to_str().unwrap_or_default()
                 ));
-            }
-            let (tablet_id, tablet_suffix) = get_id_and_suffix_from_path(path);
-            self.create_tablet(tablet_id, tablet_suffix)
+            };
+
+            self.inner.create_tablet(path)
```

Comment:
What if `options.create_new` is true?

Reply:
We should return error if exists even in open_tablet_raw?

Reply:
Yes, I think everything checked in `open_tablet` should be also checked in `open_tablet_raw`. An simple implementation is let `open_tablet` call `open_tablet_raw`.